File: Regression/VAR_1/R/filter_fcn.R
filter_fcn <- function(prmtr){
  
  #prmtr = trans(prmtr_in)
  prmtr = trans(prmtr)
  
  phi_y1 = prmtr[1]
  phi_y2 = prmtr[2]
  phi_yx = prmtr[3]
  
  phi_h1 = prmtr[4]
  phi_h2 = prmtr[5]
  phi_hx = prmtr[6]
  
  sig_nyy = prmtr[7]^2 # s.e. of credit permanent component
  sig_eyy = prmtr[8]^2 # s.e. of credit transitory component
  sig_nhh = prmtr[9]^2 # s.e. of HP permanent component
  sig_ehh = prmtr[10]^2 # s.e. of HP transitory component
  
  sig_eyeh = prmtr[11]*sqrt(sig_ehh*sig_eyy) # corr of cross transitory components
  
  F = matrix(0,6,6) # Transition matrix
  F[1,] = c(1,0,0,0,0,0)
  F[2,] = c(0,phi_y1,phi_y2,0,phi_yx,0)
  F[3,] = c(0,0,0,0,0,0)
  F[4,] = c(0,0,0,1,0,0)
  F[5,] = c(0,phi_hx,0,0,phi_h1,phi_h2)
  F[6,] = c(0,0,0,0,0,0)
  
  Fstar = F[-c(1,4), -c(1,4)]
  
  muvec = matrix(c(0,0,0,0,0,0),6,1) # Drift vector
  
  H = matrix(0,2,6)
  H[1,] = c(1,1,0,0,0,0)
  H[2,] = c(0,0,0,1,1,0)
  
  Q = matrix(0,6,6) # Cov matrix
  Q[1,] = c(sig_nyy,0,0,0,0,0)
  Q[2,] = c(0,sig_eyy,0,0,sig_eyeh,0)
  Q[3,] = c(0, 0, 0, 0, 0, 0)
  Q[4,] = c(0, 0, 0, sig_nhh, 0, 0)
  Q[5,] = c(0,sig_eyeh,0,0,sig_ehh,0)
  Q[6,] = c(0, 0,0, 0, 0, 0)
  
  Qstar = Q[-c(1,4), -c(1,4)]
  
  A = matrix(c(0,0),2,1)
  
  beta_ll = matrix(c(prior[1],0,0,prior[2] ,0,0),6,1) 
  #starting value of beta_ll is the first row of available data
  
  vecQstar = matrix(Qstar,ncol = 1)
  vecP_ll = solve(diag(16) - Fstar%x%Fstar)%*%vecQstar
  
  #solve(a,b) is to find x for a*x =b
  #b is a Unit matrix if not specified 
  #solve() is used to produce invert of a variable
  
  
  # Var matrix of initial state vector
  #P_ll = P_ll_prior
  P_ll = matrix(0,6,6)
  P_ll[1,] = c(prior[3],0,0,0,0,0)
  P_ll[2,] = c(0,vecP_ll[1,1],0,0,vecP_ll[3,1],0)
  P_ll[3,] = c(0,0,0,0,0,0)
  P_ll[4,] = c(0,0,0,prior[4],0,0)
  P_ll[5,] = c(0,vecP_ll[9,1],0,0,vecP_ll[11,1],0)
  P_ll[6,] = c(0,0,0,0,0,0)
  #force the initial matrix to be positive definite
  #P_ll = nearPD(P_ll,maxit=10)$mat 
  
  beta_mat = matrix(0,T,6)
  fcst_mat = matrix(0,T,2)
  
  for(j_iter in 1:T){
    beta_tl = muvec + F%*%beta_ll
    P_tl = F%*%P_ll%*%t(F) + Q
    
    vt = as.numeric(matrix(c(y[j_iter,1:2]),2,1)) - H%*%beta_tl - A # Prediction error
    ft = H%*%P_tl%*%t(H) # Variance of forecast error
    
    beta_tt = beta_tl + P_tl%*%t(H)%*%solve(ft)%*%vt
    P_tt = P_tl - P_tl%*%t(H)%*%solve(ft)%*%H%*%P_tl
    
    beta_ll = beta_tt
    P_ll = P_tt
    
    fcst_mat[j_iter,] = t(vt)
    beta_mat[j_iter,] = t(beta_tt)
  }
  
  return(list(beta_mat,fcst_mat))
}

File: Regression/VAR_1/Matlab/filter_fcn.m
function [beta_mat,fcst_mat] = filter_fcn(prmtr,y,T,START,prior)

%========================================================================%

  %Transform hyperparamters to impose constraints
     
  prmtr = trans(prmtr);
    
  phi_y1 = prmtr(1);
  phi_y2 = 0;
  phi_yx = 0;
  
  phi_h1 = prmtr(2);
  phi_h2 = 0;
  phi_hx = 0;
  
  sig_nyy = prmtr(3)^2; % s.e. of HP permanent component
  sig_eyy = prmtr(4)^2; % s.e. of credit permanent component
  sig_nhh = prmtr(5)^2; % s.e. of the HP AR component
  sig_ehh = prmtr(6)^2; % s.e. of the credit AR component
  sig_eyeh = prmtr(7)*sqrt(sig_eyy*sig_ehh);
  sig_nynh = prmtr(8)*sqrt(sig_nyy*sig_nhh);

  mu = 0;
  
    F = [1,0,0,0; %Transition matrix
         0,phi_y1,0,0;         
         0,0,1,0;
         0,0,0,phi_h1];

    Fstar = [phi_y1,0;
             0,phi_h1]; %Transition matrix of I(0) part, no trends;
    
    H = [1,1,0,0; %Measurement equation
        0,0,1,1];

    Q = [sig_nyy,0,sig_nynh,0; %Cov matrix
         0,sig_eyy,0,sig_eyeh;
         sig_nynh, 0,  sig_nhh, 0;
         0,sig_eyeh,sig_ehh,0];

    Qstar = [sig_eyy,sig_eyeh; %Cov matrix of I(0) part
             sig_eyeh,sig_ehh];

    A = [0;0];

    beta_ll = [prior(1),0,prior(2),0]'; %Starting values

    vecQstar = reshape(Qstar,[numel(Qstar),1]);
    vecP_ll = inv(eye(4) - kron(Fstar,Fstar))*vecQstar;
    
    %Var matrix of initial state vector
    P_ll = [prior(3),0,0,0;
            0,vecP_ll(1,1),0,vecP_ll(2,1);
            0,0,prior(4),0;
            0,vecP_ll(3,1),0,vecP_ll(4,1)];
    
    muvec = [0,0,mu,0]'; %Drift vector
    
%========================================================================%
               
    beta_mat = zeros(T,4);
    fcst_mat = zeros(T,2);
    
    for j_iter = 1:T
        beta_tl  = muvec + F*beta_ll;
        P_tl  =  F*P_ll*F' + Q;

        vt = y(j_iter,1:2)' - H*beta_tl - A; %Prediction error

        ft =  H*P_tl*H'; %Variance of forecast error

        beta_tt = beta_tl + P_tl*H'*inv(ft)*vt;
        P_tt = P_tl - P_tl*H'*inv(ft)*H*P_tl;

        beta_ll = beta_tt;
        P_ll = P_tt;
        
        fcst_mat(j_iter,:) = vt';
        beta_mat(j_iter,:) = beta_tt';
    end
        
end
        